pythonscript/_godot_io.pxi

```diff
@@ -0,0 +1,163 @@
+)
+from godot._hazmat.gdnative_api_struct cimport (
+    godot_string,
+    godot_string_name,
+    godot_bool,
+    godot_array,
+    godot_pool_string_array,
+    godot_object,
+    godot_variant,
+    godot_variant_call_error,
+    godot_method_rpc_mode,
+    godot_pluginscript_script_data,
+    godot_pluginscript_instance_data,
+    godot_variant_call_error_error,
+    godot_variant_type
+)
+
+class GodotIOStream(RawIOBase):
+
+    def __init__(self, godot_print_func):
+        self.buffer = ""
+        self.godot_print_func = godot_print_func
+
+    def write(self, b):
+        self.buffer += b
+        if "\n" in self.buffer:
+            *to_print, self.buffer = self.buffer.split("\n")
+            self.godot_print_func("\n".join(to_print))
+
+    def flush(self):
+        if self.buffer:
+            self.godot_print_func(self.buffer)
+            self.buffer = ""
+
+class GodotIO:
+
+    _godot_stdout_io = None
+    _godot_stderr_io = None
+    _builtin_print = None
+    _traceback_print_exception = None
+
+    @staticmethod
+    def godot_print_pystr(pystr):
+        """
+            Receives a python string (pystr), convert to a godot string, and print using the godot print function.
+        """
+        cdef godot_string gdstr
+        pyobj_to_godot_string(pystr, &gdstr)
+        with nogil:
+            gdapi10.godot_print(&gdstr)
+            gdapi10.godot_string_destroy(&gdstr)
+    
+    @staticmethod
+    def godot_print_error_pystr(pystr, lineno=None, filename=None, name=None):
+        """
+            Receives a python string (pystr), convert to char*, and print using the godot_print_error function.
+            Also tries to get exception information such as, file name, line numer, method name, etc
+            and pass that along to godot_print_error
+        """
+        import traceback
+
+        # we are printing an error message, so we must avoid other errors at all costs,
+        # otherwise the user may never see the error message printed, making debugging a living hell
+        try:
+            if lineno is not None and filename is not None and name is not None:
+                # don't try to get exception info if user provided the details.
+                exc_info = sys.exc_info()
+                tb = exc_info[2]
+                if tb:
+                    tblist = traceback.extract_tb(tb)
+                    if len(tblist) > 0:
+                        lineno = tblist[-1].lineno
+                        filename = tblist[-1].filename
+                        name = tblist[-1].name
+        except:
+            sys.__stderr__.write('Additional errors occured while printing:\n' + traceback.format_exc() + '\n')
+        
+        # default values in case we couldn't get exception info and user have not provided those
+        lineno = lineno or 0
+        filename = filename or '<UNKNOWN>'
+        name = name or '<UNKNOWN>'
+        
+        # TODO: how to handle different encodings?
```

Comment:
Do we really need different encoding ? I would say Godot should be responsible to display the log to the user so only it should worry about what encoding the OS supports.
From what I remember, Godot only supports UCS2 (UCS4 on non windows platforms), so we should only worry to encode to those format.
That's what is done currently, to convert Python string to Godot string, see https://github.com/touilleMan/godot-python/blob/292d39b899b5c8f4dac8971dd1fe4fb4ff3baa5c/pythonscript/godot/_hazmat/conversion.pxd#L16-L34

I guess you should use the `pyobj_to_godot_string` function, then use the godot api to access the underlying char* buffer of the resulting godot_string

Reply:
I tried doing this, and the way to it is:

create godot_string with `pyobj_to_godot_string`
convert godot string to godot_char_string with `gdapi10.godot_string_utf8`
convert to char* using `gdapi10.godot_char_string_get_data`
deallocate godot_char_string
deallocate godot_string

this end up being a lot of code and looking horrible, to in the end do the same thing we were doing anyway: convert to utf8. I'm gonna stick with this implementation for now.